### PR TITLE
feat(bailian): add native support for Alibaba Cloud Bailian (百炼)

### PR DESCRIPTION
**Pull request by Barsit**
---
  ## Summary

  Adds **Alibaba Cloud Bailian (百炼)** (`dashscope.aliyuncs.com`) as a primary provider. Setting `BAILIAN_API_KEY` is
  the whole setup (`BAILIAN_BASE_URL` optional for region switching).

  ### Provider

  - New `internal/providers/bailian` package, type `bailian`, default base URL
  `https://dashscope.aliyuncs.com/compatible-mode/v1`, Bearer auth.
  - Chat completions, streaming, ListModels, Responses (via chat translation), Embeddings, Files, Batches, and
  Passthrough all delegate to `openai.CompatibleProvider`.
  - Embedding models (`text-embedding-v3`, `text-embedding-v4`) are served by the compatible-mode API but are not
  auto-discovered via `GET /v1/models` — configure via `BAILIAN_MODELS`.

  ### `max_tokens` compatibility shim

  Bailian deprecated `max_tokens` on 2026-04-20 in favor of `max_completion_tokens`. The provider transparently maps
  `max_tokens` → `max_completion_tokens` for every request. If the caller explicitly sets `max_completion_tokens`, that
  value is preserved and `max_tokens` is only stripped (not overridden).

  ### Multi-region

  Bailian is available in Beijing (default), Singapore, Frankfurt, and Hong Kong. `BAILIAN_BASE_URL` selects the region.

  ### Docs

  New `docs/providers/bailian.mdx` (`max_tokens` compatibility, multi-region base URLs, model IDs), wired into
  nav/overview; README table, `.env.template`, `config.example.yaml`, CLAUDE.md, Swagger `@description` updated.

  ## Testing

  - **31 unit tests** covering:
    - Happy paths: ChatCompletion, Streaming, ListModels, Embeddings, Responses, Passthrough, CreateBatch, GetBatch,
  ListBatches, CancelBatch, CreateFile, ListFiles, GetFile, DeleteFile, GetFileContent, GetBatchResults
    - Error paths: upstream 400/401/500 for ChatCompletion, StreamChatCompletion, Responses, Embeddings, ListModels,
  Passthrough; transport failure
    - Provider-specific: `max_tokens` → `max_completion_tokens` mapping, explicit `max_completion_tokens` preservation,
  nil/empty edge cases, request cloning, field preservation, Registration type and default URL
  - **Coverage: 91.5%**
  - **E2E:** existing generic gateway tests pass; no provider-specific e2e fixtures pollute the shared test harness
  - **Manually verified** against the live Bailian API (`dashscope.aliyuncs.com`): chat completions (non-streaming +
  SSE), `/v1/responses`, and `/v1/embeddings` all confirmed working

  🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for Alibaba Cloud Bailian (百炼) as an LLM provider, enabling chat completions, embeddings, file operations, and batch processing with Qwen models.

* **Documentation**
  * Added comprehensive provider documentation and configuration examples for Bailian setup.

### CODE REVIEW

**Comment by Barsit** [comment]
这是来自QQ邮箱的自动回复邮件。您的邮件我已收到。

File: internal/providers/bailian/bailian.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	return providers.ResponsesViaChat(ctx, p, req)
+}
+
+func (p *Provider) StreamResponses(ctx context.Context, req *core.ResponsesRequest) (io.ReadCloser, error) {
+	return providers.StreamResponsesViaChat(ctx, p, req, "bailian")
+}
+
+func (p *Provider) Embeddings(ctx context.Context, req *core.EmbeddingRequest) (*core.EmbeddingResponse, error) {
+	return p.compatible.Embeddings(ctx, req)
+}
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/bailian/bailian.go]: This was based on an earlier version of the PR description. After verifying against the live API, the
  ▎ compatible-mode /v1/embeddings endpoint does work correctly (tested with text-embedding-v3 returning a valid
  ▎ 1024-dim embedding). The docs and feature matrix have been updated to reflect this. The only caveat is that
  ▎ embedding models aren't auto-discovered from GET /v1/models, so users need
  ▎ BAILIAN_MODELS=text-embedding-v3,text-embedding-v4 to route to them.

File: internal/providers/bailian/bailian.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	if req == nil || req.MaxTokens == nil {
+		return req
+	}
+	maxTokensJSON, _ := json.Marshal(*req.MaxTokens)
+	extra, _ := core.MergeUnknownJSONFields(req.ExtraFields, map[string]json.RawMessage{
+		"max_completion_tokens": maxTokensJSON,
+	})
+	cloned := *req
+	cloned.ExtraFields = extra
+	cloned.MaxTokens = nil
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/bailian/bailian.go]: **Response to: Preserve explicit token cap**

  > *A client that sends `max_tokens: 100` and provider-native `max_completion_tokens: 200` will have its explicit
  Bailian value silently changed to 100.*

  Thank you for catching this. Fixed in commit `b3fe961`.

  `adaptBailianRequest` now checks `req.ExtraFields.Lookup("max_completion_tokens")` before applying the mapping. If the
  caller already set `max_completion_tokens` explicitly, its value is preserved — only `max_tokens` is stripped from
  the cloned request.

  **Behavior matrix:**

  | Sent by client | Behavior |
  |---|---|
  | `max_tokens: 100` | → `max_completion_tokens: 100` (auto-mapped) |
  | `max_completion_tokens: 200` | → `max_completion_tokens: 200` (pass-through) |
  | `max_tokens: 100` + `max_completion_tokens: 200` | → `max_completion_tokens: 200` (explicit wins, `max_tokens`
  stripped) |

  A unit test (`TestAdaptBailianRequest_RespectsExistingMaxCompletionTokens`) covers the mixed case and verifies that
  the explicit `max_completion_tokens=200` survives the shim unchanged.

File: tests/e2e/provider_bailian_test.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	maxTokens := 4096
+	payload := core.ChatRequest{
+		Model: "qwen3-max",
+		Messages: []core.Message{
+			{Role: "user", Content: "Hello Bailian"},
+		},
+		MaxTokens: &maxTokens,
+	}
+
+	resp := sendChatRequest(t, payload)
```
[the remark this answers is not in the document]
Comment (Barsit) [reply tests/e2e/provider_bailian_test.go]: **Response to: Exercise real provider**

  > *These Bailian E2E tests run against the global TestProvider, not the Bailian provider. The tests will still pass if  the Bailian registration or adapter is broken.*

  Valid point — these tests were misleading. They exercised the generic gateway routing layer under Bailian-themed model  names but never actually invoked `bailian.Provider`.

  Fixed in commit `b3fe961`:

  1. **Removed** `tests/e2e/provider_bailian_test.go` entirely.
  2. **Reverted** all Bailian-related changes to shared test infrastructure:
     - `tests/e2e/main_test.go` — removed `qwen3-*` and `text-embedding-v3` model entries from
  `TestProvider.ListModels`; restored `Embeddings` to return the original "unsupported" error.
     - `tests/e2e/mock_provider.go` — removed the `/embeddings` handler and `forwardEmbeddingsRequest`.
     - `tests/e2e/admin_test.go` — reverted model count assertion back to 3.

  The Bailian provider is now covered entirely by focused unit tests in `internal/providers/bailian/bailian_test.go` (17
  tests, **100% coverage**), which run against `httptest.NewServer` and exercise the real `bailian.Provider` code
  including `adaptBailianRequest`, `ChatCompletion`, `StreamChatCompletion`, `ListModels`, `Embeddings`, `Responses`,
  `Passthrough`, `CreateFile`, `CreateBatch`, etc.

  Existing generic E2E tests continue to validate the gateway pipeline (routing, auth, streaming) without polluting the
  shared test harness with provider-specific fixtures.

File: tests/e2e/main_test.go [hunk 4]
```diff
@@ -157,6 +157,9 @@ func (p *TestProvider) ListModels(ctx context.Context) (*core.ModelsResponse, er
 			{ID: "gpt-4.1", Object: "model", OwnedBy: "openai"},
 			{ID: "gpt-4", Object: "model", OwnedBy: "openai"},
 			{ID: "gpt-3.5-turbo", Object: "model", OwnedBy: "openai"},
+			{ID: "qwen3-max", Object: "model", OwnedBy: "bailian"},
+			{ID: "qwen3-flash", Object: "model", OwnedBy: "bailian"},
+			{ID: "text-embedding-v3", Object: "model", OwnedBy: "bailian"},
```
[the remark this answers is not in the document]
Comment (Barsit) [reply tests/e2e/main_test.go]: **Response to: E2E "Bailian provider" tests are routing to TestProvider, not the real bailian.Provider**

  Agreed on all points. These tests were not validating what their name claimed.

  Fixed in commit `b3fe961`:

  - **Deleted** `tests/e2e/provider_bailian_test.go` — the entire file is gone.
  - **Reverted** all shared test infrastructure changes:
    - `tests/e2e/main_test.go`: removed `qwen3-max`, `qwen3-flash`, `text-embedding-v3` from `TestProvider.ListModels`;
  `Embeddings` restored to returning the original "unsupported" error.
    - `tests/e2e/mock_provider.go`: removed the `/embeddings` handler and `forwardEmbeddingsRequest`.
    - `tests/e2e/admin_test.go`: model count assertion back to 3.

  The Bailian provider's request/response adaptation logic (including `adaptBailianRequest`, `ChatCompletion`,
  `StreamChatCompletion`, `ListModels`, `Embeddings`, `Responses`, `Passthrough`, `CreateFile`, `CreateBatch`) is
  exercised by unit tests in `internal/providers/bailian/bailian_test.go` (17 tests, 100% coverage) using
  `httptest.NewServer` routing directly through `bailian.Provider`.

  The generic E2E tests continue to cover gateway-level concerns (routing, auth, SSE streaming, error codes) without
  provider-specific fixtures leaking into the shared test harness.

File: config/config.example.yaml [hunk 5]
```diff
@@ -199,6 +199,11 @@ providers:
     type: anthropic
     api_key: "sk-ant-..."
 
+  bailian:
+    type: bailian
+    api_key: "${BAILIAN_API_KEY}"
+    # base_url: "https://dashscope.aliyuncs.com/compatible-mode/v1"
```
[the remark this answers is not in the document]
Comment (Barsit) [reply config/config.example.yaml]: Updated `config/config.example.yaml` with the correct workspace-scoped
  endpoints for Singapore, Frankfurt, and Hong Kong.

File: internal/providers/bailian/bailian_test.go [hunk 6]
```diff
@@ -0,0 +1,408 @@
+﻿package bailian
```
[the remark this answers is not in the document]
Comment (Barsit) [reply internal/providers/bailian/bailian_test.go]: Fixed in commit `b843e29`. Removed the UTF-8 BOM (`U+FEFF`) from the start of
  `internal/providers/bailian/bailian_test.go`. Verified `bailian.go` has no BOM. Tests pass.

File: tests/e2e/provider_bailian_test.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	defer closeBody(resp)
+
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	var chatResp core.ChatResponse
+	require.NoError(t, json.NewDecoder(resp.Body).Decode(&chatResp))
+
+	assert.Equal(t, "chat.completion", chatResp.Object)
+	assert.Equal(t, "qwen3-max", chatResp.Model)
+	assert.Contains(t, chatResp.Choices[0].Message.Content, "Hello Bailian")
```
[the remark this answers is not in the document]
Comment (Barsit) [reply tests/e2e/provider_bailian_test.go]: **Response: Validate embeddings requests before returning 200**

  This comment is now outdated. In a previous review iteration, `handleEmbeddings` was added to `mock_provider.go` along  with a `forwardEmbeddingsRequest` helper and Bailian model entries in `TestProvider.ListModels`.

  All of those changes have been reverted in commit `b3fe961`:

  - `tests/e2e/mock_provider.go`: back to original — no `/embeddings` handler.
  - `tests/e2e/main_test.go`: `TestProvider.Embeddings` returns the original `"test provider does not support
  embeddings"` error — no Bailian models in the mock registry.
  - `tests/e2e/provider_bailian_test.go`: deleted.

  The current state of `mock_provider.go` does not contain a `handleEmbeddings` function, so no fix is needed.

File: .env.template [hunk 8]
```diff
@@ -346,6 +346,10 @@
 # Set base URL to enable (default: http://localhost:11434/v1)
 # OLLAMA_BASE_URL=http://localhost:11434/v1
 
+# Alibaba Cloud Bailian (百炼)
+# BAILIAN_API_KEY=...
+# BAILIAN_BASE_URL=
+
```
[the remark this answers is not in the document]
Comment (Barsit) [reply .env.template]: ●**Response: Show the default base URL value for consistency**

  Fixed in commit `441b76c`. Changed `# BAILIAN_BASE_URL=` to `#
  BAILIAN_BASE_URL=https://dashscope.aliyuncs.com/compatible-mode/v1` to match the pattern used by all other providers
  in `.env.template`.

File: docs/providers/bailian.mdx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+| Files (`/v1/files`) | ✅ |
+| Batches (`/v1/batches`) | ✅ |
+| Passthrough (`/p/bailian/...`) | ✅ |
+
+<Note>
+  Embedding models (`text-embedding-v3`, `text-embedding-v4`) are served by
+  the compatible-mode API but are **not** auto-discovered from the upstream
+  `/v1/models` endpoint. Set `BAILIAN_MODELS=text-embedding-v3,text-embedding-v4`
+  or use `CONFIGURED_PROVIDER_MODELS_MODE=allowlist` to make them available.
+</Note>
```
[the remark this answers is not in the document]
Comment (Barsit) [reply docs/providers/bailian.mdx]: **Response: Mark Bailian embeddings as unsupported across all surfaces**

  I respectfully disagree with this comment. The PR description you're citing is from an early iteration before I
  verified against the live API.

  Testing against the production compatible-mode API confirms that `/v1/embeddings` **does work**:

  ```bash
  $ curl https://dashscope.aliyuncs.com/compatible-mode/v1/embeddings \
    -H "Authorization: Bearer $API_KEY" \
    -d '{"model":"text-embedding-v3","input":"Embeddings work"}'

  # → {"object":"list","data":[{"embedding":[...]}],"model":"text-embedding-v3"}

  The docs and feature matrix were updated from ❌  to ✅  after this live verification. The current bailian.mdx note
  accurately reflects the only caveat: embedding models (text-embedding-v3, text-embedding-v4) aren't returned by GET
  /v1/models, so they require manual configuration via BAILIAN_MODELS.

  The PR description has also been updated to reflect this finding. No change needed.
  ```

**Comment by Barsit** [comment]
@coderabbitai resume

**Comment by SantiagoDePolonia** [comment]
@Barsit Great job! Thank you!

Please:
1. Fix the lint issue.
2. Merge main and fix conflicts.

Once done, I'll merge this!
